src/live_restore/live_restore_state.c

```
@@ -0,0 +1,362 @@
+ *     can restart on completion of live restore.
+ */
+static void
+__live_restore_report_state_to_application(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_STATE state)
+{
+    switch (state) {
+    case WTI_LIVE_RESTORE_STATE_NONE:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_INIT);
+        break;
+    case WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION:
+    case WTI_LIVE_RESTORE_STATE_CLEAN_UP:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_IN_PROGRESS);
+        break;
+    case WTI_LIVE_RESTORE_STATE_COMPLETE:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_COMPLETE);
+        break;
+    }
+}
+
+/*
+ * __wti_live_restore_set_state --
+ *     Update the live restore state in memory and persist it to the turtle file.
+ */
+int
+__wti_live_restore_set_state(
+  WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_FS *lr_fs, WTI_LIVE_RESTORE_STATE new_state)
+{
+    WT_DECL_RET;
+
+    bool reentrant = __wt_spin_owned(session, &lr_fs->state_lock);
+    if (!reentrant)
+        __wt_spin_lock(session, &lr_fs->state_lock);
+
+    /*
+     * Validity checking. There is a defined transition of states and we should never skip or repeat
+     * a state.
+     */
+    switch (new_state) {
+    case WTI_LIVE_RESTORE_STATE_NONE:
+        /*  We should never transition to NONE. This is a placeholder when state is not set. */
+        WT_ASSERT_ALWAYS(session, false, "Attempting to set Live Restore state to NONE!");
+        break;
+    case WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION:
+        WT_ASSERT(session, lr_fs->state == WTI_LIVE_RESTORE_STATE_NONE);
+        break;
+    case WTI_LIVE_RESTORE_STATE_CLEAN_UP:
+        WT_ASSERT(session, lr_fs->state == WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION);
+        break;
+    case WTI_LIVE_RESTORE_STATE_COMPLETE:
+        WT_ASSERT(session, lr_fs->state == WTI_LIVE_RESTORE_STATE_CLEAN_UP);
+        break;
```

Do you want a panic `default` here and all the other switch statements. If someone adds a new state someday and forgets to add a case in one of these, then it would be detected, hopefully in testing.

---

I've defined `WTI_LIVE_RESTORE_STATE` as an enum for this purpose since `DENABLE_STRICT` turns on `Wswitch-enum` and the compiler catches unhandled cases. I think this'll be more robust as `default` would hide an unhandled case if we added a new state in future.